research/train.py

```diff
@@ -60,18 +106,35 @@ def train(cfg: DictConfig) -> None:
     # Determine device (GPU, CPU, etc.)
     device = "cuda" if torch.cuda.is_available() else "cpu"
 
+    # Model
+    model = get_network(cfg)
+
     # Data Loader
     train_loader, val_loader = get_dataloaders(cfg, num_workers=cfg.data_loader_workers)
 
     # Training loop logic
-    trainer = create_supervised_trainer(cfg, device=device)
+    trainer = create_supervised_trainer(model, cfg, device=device)
 
+    # Evaluation loop logic
+    evaluator = create_supervised_evaluator(model, cfg, device=device)
+
+    ########################################################################
     # Callbacks
+    ########################################################################
     trainer.add_event_handler(Events.STARTED, lambda engine: print("Start training"))
-    # trainer.add_event_handler(Events.EPOCH_COMPLETED, log_training_results)
-    trainer.add_event_handler(Events.ITERATION_COMPLETED, log_training_loss)
 
-    # train()
+    # When epoch completes, run evaluator engine on val_loader, then log ["accuracy", "nll"] metrics.
+    trainer.add_event_handler(
+        Events.EPOCH_COMPLETED,
+        _lf_val(log_engine_metrics_stdout, evaluator, val_loader, ["accuracy", "nll"]),
+    )
+
+    # When batch completes, log train_engine nll output for batch.
+    trainer.add_event_handler(
```

Comment:
Are there more custom events such as event_handler after every N iterations? Printing every single iteration can obscure the training data and when you train for many iterations you might onyl want to print to console every N iterations, but log to file every iteration

Reply:
Yes! You can fire the event every `n` epochs using the `every` named arg: